File: templates/python/v1/builder/http_request_builder.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict

DEFAULT_METHOD = "GET"
DEFAULT_BODY = ""
DEFAULT_TIMEOUT_MS = 5000


@dataclass(frozen=True, slots=True)
class HttpRequest:
    method: str
    url: str
    headers: Dict[str, str]
    body: str
    timeout_ms: int

    def summary(self) -> str:
        return f"{self.method} {self.url} (headers={len(self.headers)}, timeoutMs={self.timeout_ms})"


@dataclass(slots=True)
class HttpRequestBuilder:
    method: str = DEFAULT_METHOD
    url: str | None = None
    headers: Dict[str, str] = field(default_factory=dict)
    body: str = DEFAULT_BODY
    timeout_ms: int = DEFAULT_TIMEOUT_MS

    def set_url(self, url: str) -> "HttpRequestBuilder":
        self.url = url
        return self

    def set_method(self, method: str) -> "HttpRequestBuilder":
        self.method = method
        return self

    def add_header(self, key: str, value: str) -> "HttpRequestBuilder":
        self.headers[key] = value
        return self

    def set_body(self, body: str) -> "HttpRequestBuilder":
        self.body = body
        return self

    def set_timeout_ms(self, timeout_ms: int) -> "HttpRequestBuilder":
        self.timeout_ms = timeout_ms
        return self

    def build(self) -> HttpRequest:
        if not self.url:
            raise ValueError("url is required")
        return HttpRequest(
            method=self.method,
            url=self.url,
            headers=dict(self.headers),
            body=self.body,
            timeout_ms=self.timeout_ms,
        )

File: templates/javascript/v1/builder/httpRequestBuilder.js
export class HttpRequest {
  constructor({ method, url, headers, body, timeoutMs }) {
    this.method = method;
    this.url = url;
    this.headers = headers;
    this.body = body;
    this.timeoutMs = timeoutMs;
  }

  summary() {
    return `${this.method} ${this.url} (headers=${Object.keys(this.headers).length}, timeoutMs=${this.timeoutMs})`;
  }
}

export class HttpRequestBuilder {
  constructor() {
    this.method = "GET";
    this.url = null;
    this.headers = {};
    this.body = "";
    this.timeoutMs = 5000;
  }

  setUrl(url) {
    this.url = url;
    return this;
  }

  setMethod(method) {
    this.method = method;
    return this;
  }

  addHeader(key, value) {
    this.headers[key] = value;
    return this;
  }

  setBody(body) {
    this.body = body;
    return this;
  }

  setTimeoutMs(timeoutMs) {
    this.timeoutMs = timeoutMs;
    return this;
  }

  build() {
    if (!this.url) {
      throw new Error("url is required");
    }
    return new HttpRequest({
      method: this.method,
      url: this.url,
      headers: { ...this.headers },
      body: this.body,
      timeoutMs: this.timeoutMs
    });
  }
}
